src/tir/schedule/primitive/fuse_split.cc

```diff
@@ -0,0 +1,483 @@
+      : vmap_(vmap), opaque_blocks_(opaque_blocks) {}
+
+ private:
+  PrimExpr VisitExpr_(const VarNode* op) final {
+    Var var = GetRef<Var>(op);
+    Optional<PrimExpr> ret = vmap_(var);
+    if (ret.defined()) {
+      return ret.value();
+    } else {
+      return std::move(var);
+    }
+  }
+
+  Stmt VisitStmt_(const BlockRealizeNode* op) final {
+    Stmt res = StmtMutator::VisitStmt_(op);
+    if (op->block->iter_vars.empty()) {
+      const BlockRealizeNode* realize = res.as<BlockRealizeNode>();
+      opaque_blocks_->Set(op->block, realize->block);
+    }
+    return res;
+  }
+
+  /*! \brief The substitute function */
+  std::function<Optional<PrimExpr>(const Var&)> vmap_;
+  /*! \brief The reuse mapping */
+  Map<Block, Block>* opaque_blocks_;
+};
+
+Stmt SubstituteAndCollectOpaqueBlock(Stmt stmt, Map<Block, Block>* opaque_blocks,
+                                     std::function<Optional<PrimExpr>(const Var&)> vmap) {
+  return IRSubstituteAndCollectOpaqueBlock(vmap, opaque_blocks)(std::move(stmt));
+}
+
+/*! \brief Simplify the binding of block realize and update the opaque block reuse mapping*/
+class BlockRealizeRewriter : public StmtExprMutator {
+ public:
+  explicit BlockRealizeRewriter(
+      const std::unordered_map<Var, Range, ObjectPtrHash, ObjectPtrEqual>& loop_map,
+      Map<Block, Block>* opaque_blocks)
+      : opaque_blocks_(opaque_blocks) {
+    loop_map_.insert(loop_map.begin(), loop_map.end());
+  }
+
+ private:
+  Stmt VisitStmt_(const ForNode* op) final {
+    loop_map_[op->loop_var] = Range::FromMinExtent(op->min, op->extent);
+    Stmt res = StmtMutator::VisitStmt_(op);
+    loop_map_.erase(op->loop_var);
+    return res;
+  }
+
+  Stmt VisitStmt_(const BlockRealizeNode* op) final {
+    // skip opaque block and update mapping
+    if (op->iter_values.empty()) {
+      Stmt res = StmtMutator::VisitStmt_(op);
+      const BlockRealizeNode* realize = res.as<BlockRealizeNode>();
+      for (const std::pair<Block, Block>& entry : *opaque_blocks_) {
+        if (entry.second.same_as(op->block)) {
+          opaque_blocks_->Set(entry.first, realize->block);
+          break;
+        }
+      }
+      return res;
+    }
+    auto v = arith::IterMapSimplify(op->iter_values, loop_map_, op->predicate, false);
+    if (v.same_as(op->iter_values)) {
+      return GetRef<Stmt>(op);
+    } else {
+      auto n = CopyOnWrite(op);
+      n->iter_values = std::move(v);
+      return Stmt(n);
+    }
+  }
+  /*! \brief The range of loops */
+  std::unordered_map<Var, Range, ObjectPtrHash, ObjectPtrEqual> loop_map_;
+  /*! \brief The reuse mapping */
+  Map<Block, Block>* opaque_blocks_;
+};
+
+Stmt SimplifyBindings(const Stmt& stmt, const Array<StmtSRef>& loops,
+                      Map<Block, Block>* opaque_blocks) {
+  std::unordered_map<Var, Range, ObjectPtrHash, ObjectPtrEqual> loop_map;
+  for (const StmtSRef& sref : loops) {
+    const auto* loop = sref->StmtAs<ForNode>();
```

Comment:
Use the macro to enforce an ICHECK

```suggestion
    const ForNode* loop = TVM_SREF_TO_FOR(loop, sref);
```